### PR TITLE
Cuda12 rebuild

### PR DESCRIPTION
**Pull request by RaulPPelaez**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

This is my take on trying to build for CUDA 12, its just a continuation of #171.
This builds locally on my machine, but is failing on testing because it does not find the mkl .so:

```shell
import: 'torch'                                                                                                                                                              
Traceback (most recent call last):                                                                                                                                           
  File "/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1692196994653/test_tmp/run_test.py", line 2, in <module>                                                   
    import torch                                                                                                                                                             
  File "/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1692196994653/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla
cehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.11/site-packages/torch/__init__.py", line 228, in <module>              
    _load_global_deps()                                                                                                                                                      
  File "/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1692196994653/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla
cehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.11/site-packages/torch/__init__.py", line 187, in _load_global_deps     
    raise err
  File "/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1692196994653/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla
cehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.11/site-packages/torch/__init__.py", line 168, in _load_global_deps
    ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
  File "/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1692196994653/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla
cehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.11/ctypes/__init__.py", line 376, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: libmkl_intel_lp64.so.1: cannot open shared object file: No such file or directory
```

### CODE REVIEW

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

File: recipe/meta.yaml [hunk 1]
```diff
@@ -79,7 +79,9 @@ outputs:
         - requests
         - future
         - six
-        - mkl-devel {{ mkl }}   # [x86]
+        - mkl==2021.4.0         # [x86]
+        - mkl-include==2021.4.0 # [x86]
+        - mkl-devel==2021.4.0   # [x86]-
```
Comment (jakirkham) [inline recipe/meta.yaml:84]: Maybe these pins could be moved to `conda_build_config.yaml`. Then when resolving upstream changes would prefer upstream's changes (then could readd the `-devel` part). IIUC `-include` comes with `-devel`

https://github.com/conda-forge/pytorch-cpu-feedstock/blob/229b44730ae4e9b1cb144910f0a3a25cee0e7391/recipe/conda_build_config.yaml#L10-L13

Comment (h-vetinari) [reply recipe/meta.yaml:84]: They're also really old (>2 years). Do we really need to pin mkl that tightly?

**Comment by h-vetinari** [comment]
> This is my take on trying to build for CUDA 12, its just a continuation of #171.
> This builds locally on my machine, but is failing on testing because it does not find the mkl .so:

Thanks for picking this up @RaulPPelaez!

**Comment by h-vetinari** [comment]
The PR will also have to be rebased, or at least merged with `main`.

File: recipe/meta.yaml (hunk 1 above)
Comment (jakirkham) [reply recipe/meta.yaml:84]: Yeah am wondering the same. It also deviates from [the current global pinnings]( https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/023ae6b4004f90fb80116d7010ba5348be75757e/recipe/conda_build_config.yaml#L556-L559 )

Maybe this was needed to get the build working, in which case there probably needs to be some investigation into how to bump the versions (and fix whatever issue occurs when using more recent versions)

Comment (RaulPPelaez) [reply recipe/meta.yaml:84]: See this comment for the rationale: https://github.com/conda-forge/pytorch-cpu-feedstock/pull/172#issuecomment-1649225752

**Comment by RaulPPelaez** [comment]
I am having some success building locally. However, I am having trouble merging this with main, a lot of conflicts arise in files automatically created by the bot and I have no idea how to solve them.
I would like to somehow reopen this PR starting from the current main but I do not know how to instruct conda-smithy to trigger the CUDA 12 migration there.
Some help pls?

**Comment by RaulPPelaez** [comment]
Ok I think I got it, I am closing this and continuing in #182 .

**Pull request closed without merging**